### PR TITLE
Implement `torch._foreach_lerp`

### PR DESCRIPTION
**Pull request by crcrpar**
As per title.

- [ ] ~~Q: Do we want `torch._foreach_lerp.ScalarList` as well?~~
- [ ] ~~we might want to have `ATen/native/cuda/lerp.cuh` and include it in `ATen/native/cuda/Lerp.cu` and `ATen/native/cuda/ForeachTernaryOp.cu`~~

Related:
- https://github.com/pytorch/pytorch/issues/58833
- https://github.com/pytorch/pytorch/issues/71683

cc @vadimkantorov @ptrblck

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Also, unfortuately, it doesn't solve all needed foreach-needs of adam :(
e.g.:
torch._foreach_mul_(acc_deltas, rho)
torch._foreach_addcmul_(acc_deltas, deltas, deltas, value=1 - rho)

But foreach_lerp is useful for a more idiomatic EMA optimizer anyway.

File: aten/src/ATen/native/cuda/ForeachTernaryOp.cu [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  // Conditional for better numeric. This has been discussed in
+  // https://github.com/pytorch/pytorch/pull/18871
+  return (std::abs(weight_val) < 0.5)
+      ? start_val + weight * (end_val - start_val)
+      : end_val - (end_val - start_val) * (opmath_t{1} - weight_val);
+}
+
+// Ref: [GPU Pro Tip: Lerp Faster in C++](https://developer.nvidia.com/blog/lerp-faster-cuda/)
+template<>
+inline C10_HOST_DEVICE float lerp_(const float start, const float end, const float weight) {
```
Comment (vadimkantorov) [inline aten/src/ATen/native/cuda/ForeachTernaryOp.cu:32]: Why does it have underscore suffix? As far as I can see, it's not "inplace" in any sense

Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachTernaryOp.cu:32]: mimicking PowKernel.cu https://github.com/pytorch/pytorch/blob/233305a852e1cd7f319b15b5137074c9eac455f6/aten/src/ATen/native/cuda/PowKernel.cu#L23-L27

File: aten/src/ATen/native/cuda/ForeachTernaryOp.cu [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  return fma(weight, end, fma(-weight, start, start));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachTernaryOp.cu:33]: this is different than generic implementation above, and than eager implementation, do we want this divergence? Each of the branches in the generic implementation can be rewritten with 2 fmas like here, so maybe we should just do that? For fixed weight there would be no branch divergence

File: test/test_foreach.py [hunk 3]
```diff
@@ -14,7 +14,7 @@
     (instantiate_device_type_tests, dtypes, onlyCUDA, skipMeta, ops)
 from torch.testing._internal.common_methods_invocations import (
     foreach_unary_op_db, foreach_binary_op_db, foreach_pointwise_op_db, foreach_minmax_op_db,
-    foreach_reduce_op_db)
+    foreach_reduce_op_db, foreach_lerp_op_db)
```
Comment (ngimel) [inline test/test_foreach.py:17]: why do we need separate foreach_lerp_op_db, can it be part of foreach_pointwise_op_db (which, iirc, has 3 arguments).

Comment (crcrpar) [reply test/test_foreach.py:17]: the number of arguments and their pattern can be different. pointwise, addcmul and addcdiv, take three lists of tensors and an optional scalar or an optional list of scalars while the arguments of lerp can be two lists of tensors and one list of tensors and one scalar.

File: aten/src/ATen/native/cuda/ForeachTernaryOp.cu (hunk 2 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachTernaryOp.cu:33]: I wanted to make sure the precision of this fma based implementation first by comparing foreach_lerp against at::native::lerp before introducing integrating to the at::native::lerp.
I think I've overlooked some tests which confirms that parity of at::natie::lerp_cpu and at::native::lerp_cuda

Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachTernaryOp.cu:33]: so, my question is, would you mind if I update ATen/native/cuda/Lerp.cu to use this fma based logic for certain dtyps?

File: test/test_foreach.py (hunk 3 above)
Comment (ngimel) [reply test/test_foreach.py:17]: Regular ops covered by OpDB have very different arguments and yet are all tested by common tests, so I think it should be possible not to multiply dbs here

File: aten/src/ATen/native/cuda/ForeachTernaryOp.cu (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/ForeachTernaryOp.cu:33]: I don't mind fma logic, however, we should keep weight > 0.5 comparison and have 2 fma branches.

File: test/test_foreach.py (hunk 3 above)
Comment (crcrpar) [reply test/test_foreach.py:17]: After checking how the other `OpInfo` tests e.g. test_ops are doing, it seems like I should rely on `SampleInput` https://github.com/pytorch/pytorch/blob/b3206268ace6ebcb5d716ed6673876e62ef484f2/torch/testing/_internal/opinfo/core.py#L114 to better handles different arguments

**Comment by crcrpar** [comment]
This has now integrated `SampleInput` into `ForeachOpInfo` and is ready for another review

**Comment by vadimkantorov** [comment]
Another usecase for foreach_lerp is to implement explicit, manual running stats updates for a bunch of batchnorm modules (context in https://github.com/pytorch/pytorch/issues/90342#issuecomment-1341281570).

In this way, the fields running_mean/var of batchnorm modules are used to store current batch mean/var, and then the update of running_mean/var parameters is done by a separate optimizer / manually using foreach_lerp

**Comment by crcrpar** [comment]
@ngimel friendly ping

**Review by ngimel — APPROVED** [review]

**Comment by vadimkantorov** [comment]
@crcrpar Could you please then comment in https://github.com/pytorch/pytorch/issues/71683 what is implemented? Are the remaining things there the EMA optimizer? Btw now with this foreach_lerp, one could do elegant manual updates of BatchNorm stats params (with a manual call to foreach_lerp or to a separate EMA optimizer) and guard against NaN/Inf as one wishes

**Comment by vadimkantorov** [comment]
There may be another frequent idiom in optimizers:
https://github.com/pytorch/pytorch/issues/71683#issue-1111826909

that would be fixed by fused op for alpha*tensor1*tensor2 + beta*tensor3*tensor4 (for symmetricitiy): as proposed in https://github.com/pytorch/pytorch/issues/79352#issuecomment-1152946076

File: aten/src/ATen/native/ForeachOpsKernels.cpp [hunk 4]
```diff
@@ -283,6 +285,25 @@ FOREACH_POINTWISE_OP_SCALARLIST(addcmul);
 FOREACH_POINTWISE_OP_TENSOR(addcdiv);
 FOREACH_POINTWISE_OP_TENSOR(addcmul);
 
+#define FOREACH_TERNARY_OP(OP)                                                                                           \
+std::vector<Tensor> foreach_tensor_ternary_##OP##_slow(TensorList tensors1, TensorList tensors2, TensorList tensors3) {  \
+  check_foreach_api_restrictions(tensors1, tensors2, tensors3);                                                          \
+  std::vector<Tensor> result;                                                                                            \
```
Comment (Skylion007) [inline aten/src/ATen/native/ForeachOpsKernels.cpp:291]: @crcrpar Probably should have done and `result.reserve(tensors1.size());` here.
